# Linear Regression

In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [2]:
import pandas as pd
import statsmodels.formula.api as smf # multiple linear regression

In [3]:
# Load the data
url = 'https://raw.githubusercontent.com/dsahota-applied-data-analysis/data/main/Carseats.csv'

carseat_data = pd.read_csv(url)
carseat_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   400 non-null    int64  
 1   Sales        400 non-null    float64
 2   CompPrice    400 non-null    int64  
 3   Income       400 non-null    int64  
 4   Advertising  400 non-null    int64  
 5   Population   400 non-null    int64  
 6   Price        400 non-null    int64  
 7   ShelveLoc    400 non-null    object 
 8   Age          400 non-null    int64  
 9   Education    400 non-null    int64  
 10  Urban        400 non-null    object 
 11  US           400 non-null    object 
dtypes: float64(1), int64(8), object(3)
memory usage: 37.6+ KB


In [4]:
carseat_data.head()

Unnamed: 0  Sales  CompPrice  Income  Advertising  Population  Price  \
0           1   9.50        138      73           11         276    120   
1           2  11.22        111      48           16         260     83   
2           3  10.06        113      35           10         269     80   
3           4   7.40        117     100            4         466     97   
4           5   4.15        141      64            3         340    128   

  ShelveLoc  Age  Education Urban   US  
0       Bad   42         17   Yes  Yes  
1      Good   65         10   Yes  Yes  
2    Medium   59         12   Yes  Yes  
3    Medium   55         14   Yes  Yes  
4       Bad   38         13   Yes   No

## Question a)
#### Fit a multiple regression model to predict Salses using price, Urban, and US

In [5]:
# Using class notebooks as reference
sales_est = smf.ols('Sales ~ Price + Urban + US', carseat_data).fit()
sales_est.summary().tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

## Question b)
#### Provide an interpretation of each coefficient in the model. Be careful -- some of the variables in the model are qualitative!

*Qualitative variables: Urban and US *


#### **Intercept Coefficient:** 13.0435
This is the expected Sales when all other variables are 0, meaning that we expect to sell 13043 car seats if we do not consider the other variables in the model. This is the "constant" of the model.


#### **Quantitative: Urban[T.Yes] Coefficient:** -0.0219
This indicates the difference between sales when carseats are sold in Urban areas vs. non-Urban areas.
Since the sign is negative, this indicates that carseats sold in Urban areas are associated with 0.0219 decrease per unit sale than those in non-Urban areas.


#### **Quantitative: US[T.Yes] Coefficient:** 1.2006
This indicates the difference in unit sales of carseats in the US vs. outside of the US. The positive coefficient means that carseats in the US are associated with 1.2006 units higher sales than those sold elsewhere.


####  **Price Coefficient:**-0.0545
This indicates that a one-unit increase in price is associated with a 0.0545 unit (per thousand) decrease in sales of carseats since the sign is negative (sales are in thousands, so 54 unit decrease)


## Question c)
#### Write out the model in equation form, being careful to handle the quantitative variables properly.

Sales = $β_{0}$ + $β_{1}$ x Price + $β_{2}$ x Urban[Yes] + $β_{3}$ x US[Yes] + ϵ

<br>

**ϵ** : error term of the model

<br>

#### **$β_{0}$** : Intercept Coefficient, 13.0435
#### **$β_{1}$** : Urban Yes Coefficient, -0.0219
#### **$β_{2}$** : US Yes Coefficient, 1.2006
#### **$β_{3}$** : Price Coefficient, -0.0545

<br>

#### For Quantitative variables:
#### **Urban[Yes]** : 1 if Urban, 0 otherwise -> Dummy variables
#### **US[Yes]** : 1 if in US, 0 otherwise -> Dummy variables

<br>

 Sales = 13.0435 - 0.0219 x Urban[Yes] + 1.2006 x US[Yes] - 0.0545 x Price + ϵ


 This is the equation for the model. It allows you to predict the sales of carseats based on the Price, Urban statue, and US status. It handles the quantitative variables by assigning them to 1 if they apply, and 0 otherwise.

## Question d)
#### For which of the predictors can you reject the null hypothesis ${H}_0$ : ${B}_j$ = 0 with a p-value of 0.05?

P-Values for each variable (from the data above):

#### Urban[T.Yes] : 0.936
#### US[T.Yes] : 0.000
#### Price : 0.000


For Urban: Since the p-value for Urban status is greater than 0.05, we **cannot** reject the null hypothsis ${H}_0$ : ${β}_{Urban}$ = 0. This means that the status of being Urban or non-Urban does not have a statistically significant effect on Sales at the 0.05 significance level.

For US: The p-value for US status is 0.0, which is less than the p-value of the null hypothesis. We **can** reject ${H}_0$ : ${β}_{US}$ = 0, meaning that the Sales **are** significantly effected by the carseat sold in the US vs. elsewhere.

For Price: The p-value for Price is 0.0, which is less than the p-value of the null hypothsis. Thus, we **can** reject the null hypothsis ${H}_0$ : ${β}_{Price}$ = 0. This means that the Sales **are** significantly effected by the carseat price.


## Question e)
#### On the basis of your response to the previous question, fit a smaller model that only uses the predictors for which there is evidence of association with the outcome.

In [6]:
# Only model Price and US since this data has evidence of association

sales_est_small = smf.ols('Sales ~ Price + US', carseat_data).fit()

## Question f)
#### How well do the models in (a) and (e) fit the data?

## Model (a)

In [7]:
print(sales_est.summary())


                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.234
Method:                 Least Squares   F-statistic:                     41.52
Date:                Mon, 08 Apr 2024   Prob (F-statistic):           2.39e-23
Time:                        06:28:13   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1863.
Df Residuals:                     396   BIC:                             1879.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       13.0435      0.651     20.036   

## Model (e)

In [8]:
print(sales_est_small.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     62.43
Date:                Mon, 08 Apr 2024   Prob (F-statistic):           2.66e-24
Time:                        06:28:13   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1861.
Df Residuals:                     397   BIC:                             1873.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     13.0308      0.631     20.652      0.0

Both models have the same ${R}^2$ value (0.239), which indicates that neither model fits the data well as this value is pretty low.

 The Adjusted ${R}^2$ is marginally better in the smaller model where Urban status is not used as a predictor (0.235 compareed to original 0.234).

Both models are a a poor fit.

## Question g)
#### Using the model from (e), obtain 95% confidence intervals for the coefficient(s).

In [9]:
# Reference:
# From Slides:
# Standard errors can be used to compute confidence intervals.
# For linear regression, the 95% confidence interval for β1 is approxmiately: β1^ ±2·SE(β1)

#### **For US:**
#### ${\hat{β}}_2$ ±2·SE(${\hat{β}}_2$) = 1.2006 ± 2(0.259)

#### Confidence interval: ~ (0.6826, 1.7186)

#### **For Price:**
#### ${\hat{β}}_3$ ±2·SE(${\hat{β}}_3$) = -0.0545 ± 2(-0.0545)

#### Confidence interval: ~ (-0.1635, 0.0545)

In [10]:
# Can also use stats model to find confidence interval
# Reference:
# https://stackoverflow.com/questions/61292464/get-confidence-interval-from-sklearn-linear-regression-in-python

confidence_intervals = sales_est_small.conf_int(alpha=0.05)
confidence_intervals

0          1
Intercept  11.79032  14.271265
US[T.Yes]   0.69152   1.707766
Price      -0.06476  -0.044195

A confidence interval of 95% means that there is a 95% probability that the true value of β is in the given interval.

## Question h)
#### Is there evidence of outliers or high leverage observations in the model from (e)? If so, please identify the outliers and high-leverage points.

In [11]:
info_e = sales_est_small.get_influence().summary_frame() # Get influence measures
num_obs = int(info_e.shape[0])

 # Get metrics of interest

''' Studentized residual for outliers '''

sr = info_e['student_resid']
# Store the outliers
outliers = []
# Check if any studentized residuals are greater than 3
for i, res in enumerate(sr):
  if abs(res) > 3:
    outliers.append((i, res))
    print('Observation ', i,' is an outlier')


''' For finding high leverage '''

hd = info_e['hat_diag']
# Find the 'high' leverage cutoff
lev_cutoff = (2*2 + 2)/num_obs

print(f'The cutoff for a high leverage is: {lev_cutoff} \nThe following observations are high leverage points: \n')
# Store the high lev points
hlp_list = []
for idx, hlp in enumerate(hd):
  # Check if the leverage is greater than the cutoff
  if hlp > lev_cutoff:
    hlp_list.append((idx,hlp))
    print('Observation ', idx,'with value {:.4f}'.format(hlp))


#print(outliers)
print('Number of High Leverage Points: ', len(hlp_list))

' Studentized residual for outliers '

' For finding high leverage '

The cutoff for a high leverage is: 0.015 
The following observations are high leverage points: 

Observation  42 with value 0.0433
Observation  125 with value 0.0260
Observation  155 with value 0.0161
Observation  156 with value 0.0154
Observation  159 with value 0.0157
Observation  165 with value 0.0286
Observation  171 with value 0.0210
Observation  174 with value 0.0297
Observation  191 with value 0.0180
Observation  203 with value 0.0154
Observation  208 with value 0.0182
Observation  269 with value 0.0192
Observation  272 with value 0.0187
Observation  313 with value 0.0232
Observation  315 with value 0.0170
Observation  356 with value 0.0183
Observation  365 with value 0.0174
Observation  367 with value 0.0237
Observation  383 with value 0.0165
Observation  386 with value 0.0166
Number of High Leverage Points:  20


#### None of the observations are considered outliers since there are no observations that had an absolute studentized residual greater than 3.

#### 20 of the observations are high leverage points and had a leverage statistic above the computed cut-off.